js/ui/windowManager.js

```diff
@@ -1832,3 +1887,509 @@ const WindowManager = new Lang.Class({
+            return;
+        }
+
+        if (!session.windowBuilder)
+            return;
+        if (focusedWindow === session.windowBuilder.get_meta_window()) {
+            if (session.windowApp.get_meta_window() === previousFocused) {
+                this._rotateInActors.push(session.windowBuilder);
+                this._rotateOutActors.push(session.windowApp);
+                this._animate(session.windowApp,
+                              session.windowBuilder,
+                              Gtk.DirectionType.LEFT);
+                session.state = SessionState.BUILDER;
+            } else {
+                session.windowBuilder.show();
+                session.buttonBuilder.show();
+                session.windowApp.hide();
+            }
+        }
+    },
+
+    _animate : function(src, dst, direction) {
+        dst.rotation_angle_y = direction == Gtk.DirectionType.RIGHT ? -180 : 180;
+        src.rotation_angle_y = 0;
+        dst.pivot_point = new Clutter.Point({ x: 0.5, y: 0.5 });
+        src.pivot_point = new Clutter.Point({ x: 0.5, y: 0.5 });
+
+        /* We set backface culling to be enabled here so that we can
+         * smootly animate between the two windows. Without expensive
+         * vector projections, there's no way to determine whether a
+         * window's front-face is completely invisible to the user
+         * (this cannot be done just by looking at the angle of the
+         * window because the same angle may show a different visible
+         * face depending on its position in the view frustum).
+         *
+         * What we do here is enable backface culling and rotate both
+         * windows by 180degrees. The effect of this is that the front
+         * and back window will be at opposite rotations at each point
+         * in time and so the exact point at which the first window
+         * becomes invisible is the same point at which the second
+         * window becomes visible. Because no back faces are drawn
+         * there are no visible artifacts in the animation */
+        src.set_cull_back_face(true);
+        dst.set_cull_back_face(true);
+
+        src.show();
+        dst.show();
+        dst.opacity = 0;
+
+        let srcGeometry = src.meta_window.get_frame_rect();
+        let srcIsMaximized = (src.meta_window.maximized_horizontally &&
+                              src.meta_window.maximized_vertically);
+        let dstIsMaximized = (dst.meta_window.maximized_horizontally &&
+                              dst.meta_window.maximized_vertically);
+        if (!srcIsMaximized && dstIsMaximized)
+            dst.meta_window.unmaximize(Meta.MaximizeFlags.BOTH);
+        dst.meta_window.move_resize_frame(false,
+                                          srcGeometry.x,
+                                          srcGeometry.y,
+                                          srcGeometry.width,
+                                          srcGeometry.height);
```

Comment:
I think we probably don't want to call `move_resize_frame` here if the src window was maximized. Instead perhaps maximise the dst window?